tests/remotes/hdfs.py

```diff
@@ -144,3 +144,62 @@ def hdfs(hdfs_server):
     port = hdfs_server
     url = f"hdfs://127.0.0.1:{port}/{uuid.uuid4()}"
     yield HDFS(url)
+
+
+class WebHDFS(Base, URLInfo):  # pylint: disable=abstract-method
+    @contextmanager
+    def _webhdfs(self):
+        from hdfs import InsecureClient
+
+        client = InsecureClient(self.url, self.user)
+        yield client
+
+    def is_file(self):
+        with self._webhdfs() as _hdfs:
+            return _hdfs.status(self.path)["type"] == "FILE"
+
+    def is_dir(self):
+        with self._webhdfs() as _hdfs:
+            return _hdfs.status(self.path)["type"] == "DIRECTORY"
+
+    def exists(self):
+        with self._webhdfs() as _hdfs:
+            return _hdfs.status(self.path, strict=False) is not None
+
+    def mkdir(self, mode=0o777, parents=False, exist_ok=False):
+        assert mode == 0o777
+        assert parents
+        assert not exist_ok
+
+        with self._webhdfs() as _hdfs:
+            # NOTE: hdfs.makekdirs always creates parents
+            _hdfs.makedirs(self.path)
+
+    def write_bytes(self, contents):
+        with self._webhdfs() as _hdfs:
+            with _hdfs.write(self.path) as writer:
+                writer.write(contents)
+
+    def write_text(self, contents, encoding=None, errors=None):
+        if not encoding:
+            encoding = locale.getpreferredencoding(False)
+        assert errors is None
+        self.write_bytes(contents.encode(encoding))
+
+    def read_bytes(self):
+        with self._webhdfs() as _hdfs:
+            with _hdfs.read(self.path) as reader:
+                return reader.read()
+
+    def read_text(self, encoding=None, errors=None):
+        if not encoding:
+            encoding = locale.getpreferredencoding(False)
+        assert errors is None
+        return self.read_bytes().decode(encoding)
+
+
+@pytest.fixture
+def webhdfs(hdfs_server):
+    port = "50070"
```

Comment:
Hm, this doesn't look right. The line you've added to the docker compose file means that that port will be exposed but with a new dynamic number, so you need to ask docker for it. E.g. see how we do `port = docker_services.port_for("hdfs", 8020)` above. Maybe consider making `hdfs_server` fixture return a port map like {8020: ..., 50070: ...} instead of just port for 8020.

Btw, are there any python emulators/mockers for webhdfs? There are none for real hdfs (at least I'm not aware of them), so that's why we use real hadoop here in a docker container. Just wondering.

Reply:
I think I confused `webhdfs` with a different service... It should all be the same port, just different scheme. I'm now using the same port as `hdfs`.

No emulators or mockers that I am aware unfortunately.